visualization/app/codeCharta/codeCharta.service.ts

```diff
@@ -67,7 +66,55 @@ export class CodeChartaService {
 	private addFile(file: NameDataPair) {
 		const ccFile = getCCFile(file)
 		NodeDecorator.decorateMapWithPathAttribute(ccFile)
+		const currentFileChecksum = ccFile.fileMeta.fileChecksum
+		let currentFileName = ccFile.fileMeta.fileName
+		const storedFileNames = new Map(this.fileStates.map(file => [file.file.fileMeta.fileName, file.file.fileMeta.fileChecksum]))
+		const storedFileChecksums = new Map(
+			this.fileStates.map((file, index) => [file.file.fileMeta.fileChecksum, index] as [string, number])
+		)
+		const isDuplicate = storedFileChecksums.has(currentFileChecksum)
+
+		if (storedFileNames.has(currentFileName)) {
+			currentFileName = this.findFileName(currentFileName, isDuplicate, storedFileNames, currentFileChecksum)
+			ccFile.fileMeta.fileName = currentFileName
+		}
+		if (isDuplicate) {
+			this.fileStates[storedFileChecksums.get(currentFileChecksum)].file.fileMeta.fileName = currentFileName
+			this.recentFiles[0] = currentFileName
+			this.recentFiles.push(currentFileName)
+			return
+		}
+
 		this.fileStates.push({ file: ccFile, selectedAs: FileSelectionState.None })
+		this.recentFiles.push(currentFileName)
+	}
+
+	private findFileName(currentFileName: string, isDuplicate: boolean, storedFileNames: Map<string, string>, currentFileChecksum: string) {
+		if (storedFileNames.get(currentFileName) === currentFileChecksum) {
+			return currentFileName
+		}
+
+		let nameFound = false
+		let fileNameOccurrence = 1
+		let newFileName = currentFileName
+		while (!nameFound) {
+			newFileName = currentFileName.replace(currentFileName, `${currentFileName.split(".")[0]}_${fileNameOccurrence}.cc.json`)
+			const endOfNameIndex = currentFileName.indexOf(".")
+			newFileName =
+				endOfNameIndex >= 0
+					? [currentFileName.slice(0, endOfNameIndex), "_", fileNameOccurrence, currentFileName.slice(endOfNameIndex)].join("")
+					: `${currentFileName}_${fileNameOccurrence}`
+			// resolve if uploaded file has identical checksum and different already occurring name
+			if (isDuplicate && storedFileNames.has(newFileName) && storedFileNames.get(newFileName) === currentFileChecksum) {
+				nameFound = true
+			}
+			if (!storedFileNames.has(newFileName)) {
+				nameFound = true
```

Comment:
I think you can `return newFileName` immediately, in those cases (nameFounde = true). Then, the wrapping `else` is redundant and can be removed.

Reply:
I didn't change that. Otherwise you would have `while(true)`. But I agree that we can delete `else`